src/v/cluster/tx_coordinator_mapper.h

```diff
@@ -57,18 +58,39 @@ class tx_coordinator_mapper {
 public:
     explicit tx_coordinator_mapper(
       ss::sharded<cluster::metadata_cache>& md,
+      ss::sharded<cluster::partition_manager>& pm,
       model::topic_namespace group_topic)
       : _md(md)
+      , _pm(pm)
       , _tp_ns(std::move(group_topic)) {}
 
     std::optional<model::ntp>
-    ntp_for(const kafka::transactional_id& tx_id) const {
+    ntp_for(const kafka::transactional_id& tx_id, model::node_id self) const {
         auto cfg = _md.local().get_topic_cfg(_tp_ns);
         if (!cfg) {
             // Transaction coordinator topic not exist in cache
             return std::nullopt;
         }
         int32_t partitions_amount = cfg->partition_count;
+        for (auto p = 0; p < partitions_amount; ++p) {
+            auto tm_stm_partition = model::ntp(_tp_ns.ns, _tp_ns.tp, p);
+            auto leader = _md.local().get_leader_id(tm_stm_partition);
+            if (!leader || leader.value() != self) {
+                continue;
+            }
+            auto partition = _pm.local().get(tm_stm_partition);
+            if (!partition) {
+                continue;
+            }
+            auto stm = partition->tm_stm();
+            if (!stm) {
+                continue;
+            }
+            if (stm->check_tx_membership(tx_id)) {
```

Comment:
we should sync before checking any stm state

Reply:
If you check existing `sync()` use cases you'll see that they all happen from the same shard (core) which host stm. For example, `init_tm_tx_locally` uses `container().invoke_on` to route the control flow to `do_init_tm_tx` which in its turn does the `sync`. IMHO the best course of actions is to introduce a new method `bool hosts(partition_id, kafka::transactional_id)` to `tx_gateway_frontend` and follow the almost usual drill with `hosts` / `do_hosts`.